packages/gatsby-plugin-image/src/components/gatsby-image.browser.tsx

```diff
@@ -50,171 +53,233 @@ export interface IGatsbyImageData {
-    width,
-    height,
-    layout
-  )
+  /**
+   * Choose if setupIntersectionObserver should use the image cache or not.
+   */
+  _setupIntersectionObserver(useCache = true): void {
+    import(`./intersection-observer`).then(({ createIntersectionObserver }) => {
+      const intersectionObserver = createIntersectionObserver(() => {
+        if (this.root.current) {
+          const cacheKey = JSON.stringify(this.props.image.images)
+          this.props.onStartLoad?.({
+            wasCached: useCache && hasImageLoaded(cacheKey),
+          })
+          this.setState({
+            isLoading: true,
+            isLoaded: useCache && hasImageLoaded(cacheKey),
+          })
+        }
+      })
+
+      if (this.root.current) {
+        // @ts-ignore - hello
+        this.unobserveRef = intersectionObserver(this.root)
+      }
+    })
+  }
+
+  shouldComponentUpdate(nextProps, nextState): boolean {
+    let hasChanged = false
+
+    // this check mostly means people do not have the correct ref checks in place, we want to reset some state to suppport loading effects
+    if (this.props.image.images !== nextProps.image.images) {
+      // reset state, we'll rely on intersection observer to reload
+      if (this.unobserveRef) {
+        // unregister intersectionObserver
+        this.unobserveRef(this.root)
+
+        // // on unmount, make sure we cleanup
+        if (this.hydrated.current && this.lazyHydrator) {
+          render(null, this.root.current)
+        }
+      }
+
+      this.setState(
+        {
+          isLoading: false,
+          isLoaded: false,
+        },
+        () => {
+          this._setupIntersectionObserver(false)
+        }
+      )
+
+      hasChanged = true
+    }
+
+    if (this.root.current && !hasChanged) {
+      console.log(nextState)
+      this._lazyHydrate(nextProps, nextState)
+    }
+
+    return false
+  }
 
-  useEffect((): (() => void) | undefined => {
-    if (root.current) {
-      const hasSSRHtml = root.current.querySelector(
+  componentDidMount(): void {
+    if (this.root.current) {
+      const hasSSRHtml = this.root.current.querySelector(
         `[data-gatsby-image-ssr]`
       ) as HTMLImageElement
+      const cacheKey = JSON.stringify(this.props.image.images)
 
       // when SSR and native lazyload is supported we'll do nothing ;)
       if (hasNativeLazyLoadSupport() && hasSSRHtml && global.GATSBY___IMAGE) {
-        onStartLoad?.({ wasCached: false })
+        this.props.onStartLoad?.({ wasCached: false })
 
+        // When the image is already loaded before we have hydrated, we trigger onLoad and cache the item
         if (hasSSRHtml.complete) {
-          customOnLoad?.()
-          storeImageloaded(JSON.stringify(images))
+          this.props.onLoad?.()
+          storeImageloaded(cacheKey)
         } else {
+          // eslint-disable-next-line @typescript-eslint/no-this-alias
+          const _this = this
```

Comment:
Coudl we add a comment explaining why we need to do this (I'm assuming it's because we can't use an arrow function in the callback, as we need access to it to unregister)

Reply:
exactly